File: web/src/lib/trip/alert-text.ts
// WSF publishes three strings per bulletin: AlertFullTitle (our `title`),
// RouteAlertText (our `text`) and BulletinText (our `body`, ingested since
// 2026-09-03). For most bulletins title and text are the same sentence typed
// twice - "Sea/Brem - ADA Alert - Chimacum #2 Elevator Out of Service" above
// an identical grey line - and the copies drift only in punctuation and
// spacing ("Edm/King- Boarding" vs "Edm/King - Boarding"), so a raw string
// compare misses them. Some bulletins DO carry the real same-day truth in
// RouteAlertText ("The 0405 VASH>FAU ... are cancelled"), which is what the
// notifier parses cancellations out of, and the body is where the substance
// of a status bulletin lives. This keeps every text that says something the
// title has not already said. The alert email applies the same rule
// (libs/wsf-core/src/wsf_core/alert_text.py) - keep the two in lockstep.

/** Case, punctuation and whitespace folded away - "Edm/King- Vessel #1" and
 * "Edm/King - Vessel #1." collapse to the same key. Keeps digits so "#1" and
 * "#2" never fold together. */
function fold(s: string): string {
  return s.toLowerCase().replace(/[^a-z0-9]+/g, "");
}

/** The texts to render under an alert's title, in order (text, then body):
 * a candidate is dropped when the title covers it, when another candidate
 * strictly contains it (the body usually restates the one-liner in full),
 * or when an earlier candidate equals it. Empty when nothing adds to the
 * title. Never drops a text that EXTENDS the title - that is where the
 * cancellations live. */
export function alertDetails(alert: {
  title: string;
  text: string | null;
  body?: string | null;
}): string[] {
  const candidates = [alert.text, alert.body].map((c) => (c ?? "").trim());
  const keys = candidates.map(fold);
  const titleKey = fold(alert.title);
  return candidates.filter((_, i) => {
    const key = keys[i]!;
    if (key.length === 0 || titleKey.includes(key)) return false;
    return !keys.some((other, j) => j !== i && other.includes(key) && (other !== key || j < i));
  });
}

File: libs/wsf-core/src/wsf_core/alert_text.py
"""What to print under an alert's title.

WSF publishes three strings per bulletin and, for most of them, two are the
same sentence typed twice: AlertFullTitle and RouteAlertText drift only in
punctuation and spacing ("Edm/King- Boarding" vs "Edm/King - Boarding").
Printing both put the Labor Day title in the owner's inbox twice (bulletin
117482, 2026-09-03). This is the one rule for every renderer - the email
here, the banner in web/src/lib/trip/alert-text.ts (kept in lockstep): fold
case, punctuation and whitespace, drop a candidate the title or another
candidate already covers, keep everything that says something new. Keep
digits so "#1" and "#2" never fold together.
"""

import re

_FOLD_RE = re.compile(r"[^a-z0-9]+")


def fold(value: str) -> str:
    return _FOLD_RE.sub("", value.lower())


def alert_details(title: str, *candidates: str | None) -> list[str]:
    """The candidate texts (in the order given) that add to the title.

    A candidate is dropped when the title covers it, when another candidate
    strictly contains it (the body usually restates the one-liner in full),
    or when an earlier candidate equals it. Never drops a text that EXTENDS
    the title - that is where cancellations live."""
    texts = [(candidate or "").strip() for candidate in candidates]
    keys = [fold(text) for text in texts]
    title_key = fold(title)
    kept: list[str] = []
    for i, (text, key) in enumerate(zip(texts, keys, strict=True)):
        if not key or key in title_key:
            continue
        if any(key in other and (other != key or j < i) for j, other in enumerate(keys) if j != i):
            continue
        kept.append(text)
    return kept
